storage/series.go

```diff
@@ -233,6 +233,80 @@ func (s *dbSeries) ReadEncoded(
 	return results, nil
 }
 
+func (s *dbSeries) FetchBlocks(ctx context.Context, starts []time.Time) []FetchBlockResult {
+	var res []FetchBlockResult
+
+	s.RLock()
+
+	for _, start := range starts {
+		if b, exists := s.blocks.GetBlockAt(start); exists {
+			stream, err := b.Stream(ctx)
+			if err != nil {
+				detailedErr := fmt.Errorf("unable to retrieve block stream for series %s time %v: %v", s.seriesID, start, err)
+				res = append(res, newFetchBlockResult(start, nil, detailedErr))
+			} else if stream != nil {
+				res = append(res, newFetchBlockResult(start, []xio.SegmentReader{stream}, nil))
+			}
+		}
+	}
+
+	if !s.buffer.Empty() {
+		bufferResults := s.buffer.FetchBlocks(ctx, starts)
+		res = append(res, bufferResults...)
+	}
+
+	s.RUnlock()
+
+	sortFetchBlockResultByTimeAscending(res)
+
+	return res
+}
+
+func (s *dbSeries) FetchBlocksMetadata(ctx context.Context, includeSizes bool) (block.DatabaseBlocksMetadata, error) {
+	var (
+		res      []block.DatabaseBlockMetadata
+		multiErr xerrors.MultiError
+	)
+
+	s.RLock()
+
+	// Iterate over the data blocks
+	blocks := s.blocks.GetAllBlocks()
+	for t, b := range blocks {
+		reader, err := b.Stream(ctx)
+		// If we failed to read some blocks, skip this block and continue to get
+		// the metadata for the rest of the blocks.
+		if err != nil {
+			multiErr = multiErr.Add(fmt.Errorf("unable to retrieve block stream for series %s time %v: %v", s.seriesID, t, err))
+			continue
+		}
+		var pSize *int64
+		if includeSizes {
+			if reader == nil {
+				// If there are no datapoints in the block, continue
+				// and don't append it to the result.
+				continue
```

Comment:
Again would be good to shortcut irregardless of `includeSizes` being true or false.